Automattic/wp-calypso
# Improve API error handling for the settings/general endpoint

On the line of "let's make WooCommerce resilient to network failures or timeouts", I've started to work on data fetching and retries. I've started with a "simple" endpoint, `settings/general`, so I can validate my approach before tackling other, more complex ones.

I've divied this PR in self-contained commits, here they are:
* https://github.com/Automattic/wp-calypso/pull/16303/commits/a15350a23b1ec80b755dcf9d656faacf47071bbc and https://github.com/Automattic/wp-calypso/pull/16303/commits/8b5209ae7270fd9c27c6958f5fe51bbb1e15863f are small cleanup commits. The `error` reducer still used the old `action = { type, payload: { ... } }` format, I converted it to the new, plain format. Also, reducers that just return the current state are pointless. I'll add them back if you feel they are helpful for readability, but I personally find them just noise :)
* In https://github.com/Automattic/wp-calypso/pull/16303/commits/b8426f4e4f6856e3bd60e1fc1fddddf3f1b1b8e0, I've fixed the `retries` logic. This will (probably) eventually be refactored to use the `data-layer`, but for now I didn't want to leave it broken (the `retries` counter wasn't being passed to the `fetch` function, so it ws always `0`). Also, I added an `ERROR` constant which can be used similarly to the `LOADING` constant in reducers & selectors, to mark that a piece of data is not being fetched, but wasn't loaded correctly either. For updating that into the reducer, I added a handler for the `WOOCOMMERCE_ERROR_SET` action. So, we have `SOMETHING_REQUEST`, `SOMETHING_REQUEST_SUCCESS` and `WOOCOMMERCE_ERROR_SET` with the `SOMETHING_REQUEST` type in its `originalAction` prop. We could also use `SOMETHING_FAILURE`, but in my view those 2 approaches seem redundant, so we should pick one or the other.
* In https://github.com/Automattic/wp-calypso/pull/16303/commits/da8d5b12ff39d46bf8459a4c14ba06f141c4c988 I created a `<QuerySettingsGeneral siteId={ siteId }>` component. The idea is that you drop it into the `render()` method of your component and it automatically fetches the settings if needed. That removes the noise of having that kind of code in the component's `componentDidMount` and `componentWillReceiveProps` lifecycle methods. In https://github.com/Automattic/wp-calypso/pull/16303/commits/d9a2cfcb94ae459a41797b66c7db2469fb3513f2 I replaced all those instances with the `<QuerySettingsGeneral>` component.
* In https://github.com/Automattic/wp-calypso/pull/16303/commits/adcf504cdf95411152c19a8813c83c627d1cee64 I improved the `<QuerySettingsGeneral>` component to add an error notice that shows if there was an error fetching. That notice has a button that can trigger a new fetch.
* In https://github.com/Automattic/wp-calypso/pull/16303/commits/14c25b7720e5d737ebca312e04750e4068270707 and https://github.com/Automattic/wp-calypso/pull/16303/commits/1ebdcce748b2f762fb2056da9d790130dac306a0 I've added logic to not keep the page in the "loading" pulsating state indefinitely if something failed. So the UI needs to have 3 states: `loaded`, `loading` and `error`. For now, the `error` state is just as if the list to show was empty (no zones / methods / locations etc). This can be improved at a later time with a custom message, this is still objetively better than having it be `loading` forever.

To test:
* Go to the Shipping Zones page (`http://calypso.localhost:3000/store/settings/shipping/[your-site]`), check that everything loads and works correctly.
* Break the `settings/general` endpoint. I went to `wp-includes/plugins/woocommerce/includes/api/class-wc-rest-setting-options-controller.php`, line `127` (the first line of the `get_items()` method) and added this inoffensive code: `$a::b; // explode`.
* Refresh the page. It will take a while (the fetch will go through 5 retries), but eventually you'll see an error notice saying that your settings couldn't be loaded.
* Click on the `Try again` button. Eventually the notice will appear again.
* Fix the `settings/general` endpoint.
* *Without* reloading the page, click the `Try again` button again. This time, everything should load and work just fine.

For the future:
* We should decide how to handle errors. This implements one possible approach, and if we go that way, se should ditch the `wc-api/error-reducer` reducer unless we have a good reason to keep it. Right now it isn't used anywhere, so it's basically dead code. It's also quite limited (only stores the last error).
* The `retries` logic should be moved to the `data-layer` eventually, or even to the `<QuerySettingsGeneral>` component.
* We're probably going to find ourselves writing a few extremely similar `<Query*>` components. That's OK for now, but eventually we could look into unifying them a bit.
